### PR TITLE
'events-async'

### PR DESCRIPTION
**Pull request by Siarhei-Balonikau**

### CODE REVIEW

File: src/modules/dirwatcher.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+  watch (path, delay) {
+    let previousStats = [];
+
+    setInterval(() => {
+      fs.readdir(path, (err, newFiles) => {
+        let newStats = this.getStats(path, newFiles);
+
+        newStats.then(currentStats => {
+          if (err) {
```
Comment (slowpokex) [inline src/modules/dirwatcher.js]: Maybe need to move error handling higher?

File: src/modules/dirwatcher.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+          }
+
+          let {deletedStats, addedStats} = this.checkFiles(previousStats, currentStats);
+
+          currentStats.sort(this.compareIno);
+
+          if (deletedStats.length !== 0) {
+            previousStats = this.deleteStats(deletedStats, previousStats);
+
+            //console.log('deleted');
```
Comment (slowpokex) [inline src/modules/dirwatcher.js]: Please remove all consoles

File: src/modules/dirwatcher.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  }
+
+  getEventEmitter() {
+    return this.eventEmitter;
+  }
+
+  watch (path, delay) {
+    let previousStats = [];
+
+    setInterval(() => {
```
Comment (slowpokex) [inline src/modules/dirwatcher.js:19]: Do you clear this interval?

File: src/modules/dirwatcher.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+            //console.log('changed');
+            this.eventEmitter.emit('changed');
+          }
+        });
+      });
+    }, delay);
+  }
+
+  getStats (path, files) {
```
Comment (slowpokex) [inline src/modules/dirwatcher.js:55]: You use env preset and you can use async/await
```
async getStats(path = '', files = []) {
  return Promise.all(files.map(filename => this.getStat(path + filename)));
}
```

File: src/modules/dirwatcher.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+      return this.getStat(fullPath);
+    });
+
+    return Promise.all(stats);
+  }
+
+  getStat (path) {
+    return new Promise((resolve, reject) => {
+      fs.stat(path, (err, stats) => {
```
Comment (slowpokex) [inline src/modules/dirwatcher.js]: Please use promisify method from util package

Comment (slowpokex) [reply src/modules/dirwatcher.js]:
```
import { promisify } from 'util';

const fsStatAsync = promisify(fs.stat);
fsStatAsync('lolkek.jpg').then().catch()
```

Comment (slowpokex) [reply src/modules/dirwatcher.js]: And so after this changes current methos will be useless

File: src/modules/dirwatcher.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+      });
+    });
+  }
+
+  compareIno (objectA, objectB) {
+    return objectA.ino - objectB.ino;
+  }
+
+  clean (array) {
+    for (var i = 0; i < array.length; i++) {
```
Comment (slowpokex) [inline src/modules/dirwatcher.js]: This method is strange, can you refactor them? And what do this method?

File: src/modules/dirwatcher.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+      if (oldFile['mtime'].getTime() !== newFiles[index]['mtime'].getTime() || oldFile['ctime'].getTime() !== newFiles[index]['ctime'].getTime()) {
+        isModified = true;
+      }
+    });
+
+    return isModified;
+  }
+
+  deleteStats(deletedStats, previousStats) {
+    return deletedStats.reduce((previousValue, delFile, index, array) => {
```
Comment (slowpokex) [inline src/modules/dirwatcher.js:116]:
```
return deletedStats.reduce((previousValue, delFile, index, array) => previousValue.filter(file => delFile.ino !== file.ino), previousStats);
```

File: src/modules/importer.js [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        result.then(data => {
+          console.log(data);
+        });
+      }
+    });
+  }
+
+  import(path) {
+    return new Promise((resolve, reject) => {
+      fs.readdir(path, (err, files) => {
```
Comment (slowpokex) [inline src/modules/importer.js:25]: Please promisify by example above

File: src/modules/importer.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+          return this.csv(csv);
+        }).then(data => {
+          resolve(data);
+        });
+      });
+    });
+  }
+
+  importSync(path) {
+    let files = fs.readdirSync(path);
```
Comment (slowpokex) [inline src/modules/importer.js:41]: Please dont create variables, if they used only once

File: src/modules/importer.js [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    let csv = this.readFilesSync(path, files);
+    let content = this.csvSync(csv);
+
+    return content;
+  }
+
+  readFiles(path, files) {
+    let content = files.map(file => {
```
Comment (slowpokex) [inline src/modules/importer.js:49]: Please short this method by example above

File: src/modules/importer.js [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+      let fullPath = path + file;
+
+      return this.readFile(fullPath);
+    });
+
+    return Promise.all(content);
+  }
+
+  readFilesSync(path, files) {
```
Comment (slowpokex) [inline src/modules/importer.js:58]: SAme please short

File: src/modules/importer.js [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+      return this.readFileSync(fullPath);
+    });
+
+    return content;
+  }
+
+  readFile(path) {
+    return new Promise((resolve, reject) => {
+      fs.readFile(path, 'utf8', (err, stats) => {
```
Comment (slowpokex) [inline src/modules/importer.js]: promisify

File: src/modules/importer.js [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+          return reject(err);
+        }
+
+        resolve(stats);
+      });
+    });
+  }
+
+  readFileSync(path) {
+    return fs.readFileSync(path, 'utf8');
```
Comment (slowpokex) [inline src/modules/importer.js:67]: utf-8 is by default - please remove this method

**Review by slowpokex — CHANGES_REQUESTED** [review]

File: src/modules/dirwatcher.js (hunk 1 above)
Comment (Siarhei-Balonikau) [reply src/modules/dirwatcher.js]: Sure, thanks)

File: src/modules/dirwatcher.js (hunk 2 above)
Comment (Siarhei-Balonikau) [reply src/modules/dirwatcher.js]: Fixed

File: src/modules/dirwatcher.js (hunk 5 above)
Comment (Siarhei-Balonikau) [reply src/modules/dirwatcher.js]: Done

File: src/modules/dirwatcher.js (hunk 7 above)
Comment (Siarhei-Balonikau) [reply src/modules/dirwatcher.js:116]: I'm not sure, that put so much logic in one row is a good idea) I deleted one useless variable 'previousValue', but generally I want to save current view. Please, tell me, if I wrong)

File: src/modules/importer.js (hunk 8 above)
Comment (Siarhei-Balonikau) [reply src/modules/importer.js:25]: In task: 'import(path): should return a promise with imported data'
If I promisify this method, how can I write this function? I don't see solution with promisify...

File: src/modules/importer.js (hunk 9 above)
Comment (Siarhei-Balonikau) [reply src/modules/importer.js:41]: I think this is improve readability... I guess this looks pretty unclear:
'return this.csvSync(this.readFilesSync(path, fs.readdirSync(path)))'
Tell me, If I wrong)

File: src/modules/importer.js (hunk 11 above)
Comment (Siarhei-Balonikau) [reply src/modules/importer.js:58]: Done

File: src/modules/importer.js (hunk 12 above)
Comment (Siarhei-Balonikau) [reply src/modules/importer.js]: Done

File: src/modules/importer.js (hunk 13 above)
Comment (Siarhei-Balonikau) [reply src/modules/importer.js:67]: Unfortunately, it returns raw buffer by default. Pls, correct me, if I wrong

File: src/modules/importer.js (hunk 10 above)
Comment (Siarhei-Balonikau) [reply src/modules/importer.js:49]: Improving readability

File: src/modules/dirwatcher.js (hunk 3 above)
Comment (Siarhei-Balonikau) [reply src/modules/dirwatcher.js:19]: Please, could you show me where I should do that? I don't have method for rejecting watcher, so I don't see place for clearing...

File: src/modules/dirwatcher.js (hunk 6 above)
Comment (Siarhei-Balonikau) [reply src/modules/dirwatcher.js]: I don't see way for refactoring, but I agree that method has unobvious behavior. So, I added comment for method.

File: src/modules/dirwatcher.js [hunk 14]
```diff
@@ -56,28 +56,17 @@ export default class DirWatcher {
...
-  }
-
   compareIno (objectA, objectB) {
     return objectA.ino - objectB.ino;
   }
 
+  /* method for clearing array from empty slots */
   clean (array) {
     for (var i = 0; i < array.length; i++) {
       if (array[i] == undefined) {
```
Comment (slowpokex) [inline src/modules/dirwatcher.js]: if (array[i]) {
}

File: src/modules/dirwatcher.js [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    return deletedStats.reduce((previousValue, delFile, index, array) => {
+      return previousValue.filter(file => delFile.ino !== file.ino);
+    }, previousStats);
+  }
+
+  addStats(addedStats, previousStats) {
+    previousStats = previousStats.concat(addedStats);
+    previousStats.sort(this.compareIno);
+
+    return previousStats;
```
Comment (slowpokex) [inline src/modules/dirwatcher.js:125]:
```
return previousStats.concat(addedStats).sort(this.compareIno);
return [...previousStats, ...addedStats].sort(this.compareIno)
```

File: src/modules/importer.js [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+        result.then(data => {
+          console.log(data);
+        });
+      }
+    });
+  }
+
+  import(path) {
+    return new Promise((resolve, reject) => {
+      fs.readdir(path, (err, files) => {
```
Comment (slowpokex) [inline src/modules/importer.js:25]:
```
const readDirAsync = promisify(fs.readdir);
return readDirAsync('lolkek/').then((files) => {
   return this.readFiles(path, files);
}).then(csv => this.csvToJSON(csv));
```

**Pull request merged**